src/SparkConduit.sol

```diff
@@ -0,0 +1,317 @@
+        // Convert the DSR a yearly APR
+        uint256 dsr      = (PotLike(pot).dsr() - RAY) * SECONDS_PER_YEAR;
+        uint256 deposits = assets[asset].totalDeposits;
+
+        return InterestData({
+            baseRate:    uint128(dsr + subsidySpread),
+            subsidyRate: uint128(dsr),
+            currentDebt: uint128(deposits),
+            targetDebt:  uint128(deposits - assets[asset].totalWithdrawals)
+        });
+    }
+
+    /// @inheritdoc ISparkConduit
+    function setRoles(address _roles) external auth {
+        roles = _roles;
+
+        emit SetRoles(_roles);
+    }
+
+    /// @inheritdoc ISparkConduit
+    function setRegistry(address _registry) external auth {
+        registry = _registry;
+
+        emit SetRegistry(_registry);
+    }
+
+    /// @inheritdoc ISparkConduit
+    function setSubsidySpread(uint256 _subsidySpread) external auth {
+        subsidySpread = _subsidySpread;
+
+        emit SetSubsidySpread(_subsidySpread);
+    }
+
+    /// @inheritdoc ISparkConduit
+    function setAssetEnabled(address asset, bool enabled) external auth {
+        assets[asset].enabled = enabled;
+        IERC20(asset).approve(address(pool), enabled ? type(uint256).max : 0);
+
+        emit SetAssetEnabled(asset, enabled);
+    }
+
+    /// @inheritdoc ISparkConduit
+    function getAssetData(address asset)
+        external view returns (bool _enabled, uint256 _totalDeposits, uint256 _totalWithdrawals)
+    {
+        uint256 liquidityIndex = pool.getReserveData(asset).liquidityIndex;
+        return (
+            assets[asset].enabled,
+            assets[asset].totalDeposits.rayMul(liquidityIndex),
+            assets[asset].totalWithdrawals.rayMul(liquidityIndex)
+        );
+    }
+
+    /// @inheritdoc ISparkConduit
+    function isAssetEnabled(address asset) external view returns (bool) {
+        return assets[asset].enabled;
+    }
+
+    /// @inheritdoc ISparkConduit
+    function getTotalDeposits(address asset) external view returns (uint256) {
+        return assets[asset].totalDeposits.rayMul(pool.getReserveData(asset).liquidityIndex);
+    }
+
+    /// @inheritdoc ISparkConduit
+    function getTotalWithdrawals(address asset) external view returns (uint256) {
+        return assets[asset].totalWithdrawals.rayMul(pool.getReserveData(asset).liquidityIndex);
+    }
+
+    /// @inheritdoc ISparkConduit
+    function getPosition(bytes32 ilk, address asset)
+        external view returns (uint256 _deposits, uint256 _withdrawals)
+    {
+        uint256 liquidityIndex = pool.getReserveData(asset).liquidityIndex;
+        Position memory position = assets[asset].positions[ilk];
+        return (
+            position.deposits.rayMul(liquidityIndex),
+            position.withdrawals.rayMul(liquidityIndex)
+        );
+    }
+
+    /// @inheritdoc ISparkConduit
+    function getDeposits(bytes32 ilk, address asset) external view returns (uint256) {
```

Comment:
I think it would be a good idea to consider `deposits` as the underlying value of the position (`shares * liquidityIndex`) and call `deposits` shares. Having `deposits` as a private mapping and `getDeposits` returning a different value could be confusing

Reply:
Yeah this is a good point. I'll rename.